File: python/tests/azure/data/postgresql/psycopg2/test_entra_id_extension.py
# Copyright (c) Microsoft. All rights reserved.

"""
Integration tests showcasing Entra ID authentication with PostgreSQL Docker instance.
These tests demonstrate token-based authentication for psycopg2.
"""

import pytest

try:
    import psycopg2
    from psycopg2.extensions import parse_dsn
except ImportError as e:
    # Provide a helpful error message if psycopg2 dependencies are missing
    raise ImportError(
        "psycopg2 dependencies are not installed. "
        "Install them with: pip install azurepg-entra[psycopg2]"
    ) from e

from testcontainers.postgres import PostgresContainer

from azurepg_entra.psycopg2.entra_connection import EntraConnection
from tests.azure.data.postgresql.test_utils import (
    TestTokenCredential,
    create_jwt_token_with_xms_mirid,
    create_valid_jwt_token,
)


@pytest.fixture(scope="module")
def postgres_container():
    """Fixture to start a PostgreSQL container for the test module."""
    with PostgresContainer("postgres:15") as container:
        yield container


@pytest.fixture(scope="module")
def connection_dsn(postgres_container) -> str:
    """Fixture to get DSN connection string from the container."""
    return (
        f"host={postgres_container.get_container_host_ip()} "
        f"port={postgres_container.get_exposed_port(5432)} "
        f"dbname={postgres_container.dbname} "
        f"user={postgres_container.username} "
        f"password={postgres_container.password}"
    )


@pytest.fixture(scope="module")
def setup_entra_users(connection_dsn):
    """Setup test users with JWT tokens as passwords."""
    # Generate JWT tokens for each user
    test_user_token = create_valid_jwt_token("test@example.com")
    managed_identity_token = create_jwt_token_with_xms_mirid(
        "/subscriptions/12345/resourcegroups/mygroup/providers/Microsoft.ManagedIdentity/userAssignedIdentities/managed-identity"
    )
    fallback_user_token = create_valid_jwt_token("fallback@example.com")
    
    setup_commands = [
        f'CREATE USER "test@example.com" WITH PASSWORD \'{test_user_token}\';',
        f'CREATE USER "managed-identity" WITH PASSWORD \'{managed_identity_token}\';',
        f'CREATE USER "fallback@example.com" WITH PASSWORD \'{fallback_user_token}\';',
        'GRANT CONNECT ON DATABASE test TO "test@example.com";',
        'GRANT CONNECT ON DATABASE test TO "managed-identity";',
        'GRANT CONNECT ON DATABASE test TO "fallback@example.com";',
        'GRANT ALL PRIVILEGES ON DATABASE test TO "test@example.com";',
        'GRANT ALL PRIVILEGES ON DATABASE test TO "managed-identity";',
        'GRANT ALL PRIVILEGES ON DATABASE test TO "fallback@example.com";',
        'GRANT ALL ON SCHEMA public TO "test@example.com";',
        'GRANT ALL ON SCHEMA public TO "managed-identity";',
        'GRANT ALL ON SCHEMA public TO "fallback@example.com";',
        'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "test@example.com";',
        'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "managed-identity";',
        'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "fallback@example.com";',
        'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "test@example.com";',
        'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "managed-identity";',
        'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "fallback@example.com";',
    ]
    
    with psycopg2.connect(connection_dsn) as conn:
        with conn.cursor() as cur:
            for sql in setup_commands:
                try:
                    cur.execute(sql)
                    conn.commit()
                except Exception:
                    # Ignore errors if user already exists
                    conn.rollback()


def assert_entra_connection_works(
    connection_dsn: str,
    token: str,
    expected_username: str
) -> None:
    """Helper to test Entra connection works end-to-end.
    
    Verifies username extraction, connection establishment, and database operations.
    """
    # Parse DSN and remove user/password
    dsn_params = parse_dsn(connection_dsn)
    dsn_params.pop('user', None)
    dsn_params.pop('password', None)
    
    # Build DSN without credentials
    base_dsn = ' '.join([f"{k}={v}" for k, v in dsn_params.items()])
    
    # Add credential
    credential = TestTokenCredential(token)
    
    # Connect using EntraConnection
    with EntraConnection(base_dsn, credential=credential) as conn:
        with conn.cursor() as cur:
            # Test basic operations
            cur.execute("SELECT current_user, current_database()")
            current_user, current_db = cur.fetchone()
            
            assert current_user == expected_username
            assert current_db == "test"


class TestPsycopg2EntraConnection:
    """Tests for psycopg2 EntraConnection."""
    
    def test_connect_with_entra_user(self, connection_dsn, setup_entra_users):
        """Showcases connecting with an Entra user using EntraConnection.
        Demonstrates: End-to-end connection with token-based authentication.
        """
        test_token = create_valid_jwt_token("test@example.com")
        assert_entra_connection_works(connection_dsn, test_token, "test@example.com")
    
    def test_connect_with_managed_identity(self, connection_dsn, setup_entra_users):
        """Showcases connecting with a managed identity using EntraConnection.
        Demonstrates: End-to-end MI authentication with token-based authentication.
        """
        xms_mirid = "/subscriptions/12345/resourcegroups/mygroup/providers/Microsoft.ManagedIdentity/userAssignedIdentities/managed-identity"
        mi_token = create_jwt_token_with_xms_mirid(xms_mirid)
        assert_entra_connection_works(connection_dsn, mi_token, "managed-identity")
    
    def test_connect_with_kwargs_override(self, connection_dsn, setup_entra_users):
        """Showcases that kwargs can override DSN parameters.
        Demonstrates: Parameter precedence (kwargs > DSN).
        """
        # Parse DSN but we'll override user via kwargs
        dsn_params = parse_dsn(connection_dsn)
        host = dsn_params['host']
        port = dsn_params['port']
        dbname = dsn_params['dbname']
        
        base_dsn = f"host={host} port={port} dbname={dbname}"
        
        # Create token for fallback user
        test_token = create_valid_jwt_token("fallback@example.com")
        credential = TestTokenCredential(test_token)
        
        # Connect - credential will extract username from token
        with EntraConnection(base_dsn, credential=credential) as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT current_user")
                current_user = cur.fetchone()[0]
                assert current_user == "fallback@example.com"

File: javascript/tests/azure/data/postgresql/test-entra-id-extension.js
/**
 * Integration tests showcasing Entra ID authentication with PostgreSQL Docker instance.
 * These tests demonstrate token-based authentication for Sequelize and node-postgres.
 */

import { describe, it, before, after } from 'mocha';
import { expect } from 'chai';
import { Sequelize } from 'sequelize';
import { GenericContainer } from 'testcontainers';
import pg from 'pg';
import { configureEntraIdAuth, getEntraTokenPassword } from '../../../../src/entra_id_extension.js';

const { Client } = pg;

/**
 * Create a base64url encoded string
 */
function createBase64UrlString(inputStr) {
    const encoded = Buffer.from(inputStr).toString('base64url');
    return encoded;
}

/**
 * Create a fake JWT token with a UPN claim
 */
function createValidJwtToken(username) {
    const header = { alg: "RS256", typ: "JWT" };
    const payload = {
        upn: username,
        iat: 1234567890,
        exp: 9999999999
    };
    
    const headerEncoded = createBase64UrlString(JSON.stringify(header));
    const payloadEncoded = createBase64UrlString(JSON.stringify(payload));
    
    return `${headerEncoded}.${payloadEncoded}.fake-signature`;
}

/**
 * Create a fake JWT token with an appid claim for managed identity
 */
function createJwtTokenWithAppId(appId) {
    const header = { alg: "RS256", typ: "JWT" };
    const payload = {
        appid: appId,
        iat: 1234567890,
        exp: 9999999999
    };
    
    const headerEncoded = createBase64UrlString(JSON.stringify(header));
    const payloadEncoded = createBase64UrlString(JSON.stringify(payload));
    
    return `${headerEncoded}.${payloadEncoded}.fake-signature`;
}

/**
 * Test token credential for mocking Azure credentials
 */
class TestTokenCredential {
    constructor(token) {
        this._token = token;
    }
    
    async getToken(scopes) {
        const expiresOnTimestamp = Date.now() + 3600000; // 1 hour from now
        return {
            token: this._token,
            expiresOnTimestamp
        };
    }
}

describe('Entra ID Extension Docker Integration Tests', function() {
    this.timeout(60000); // 60 seconds for container operations
    
    let container;
    let connectionConfig;
    
    before(async function() {
        // Start PostgreSQL container
        console.log('Starting PostgreSQL container...');
        container = await new GenericContainer('postgres:15')
            .withEnvironment({
                POSTGRES_USER: 'postgres',
                POSTGRES_PASSWORD: 'postgres',
                POSTGRES_DB: 'test'
            })
            .withExposedPorts(5432)
            .start();
        
        const host = container.getHost();
        const port = container.getMappedPort(5432);
        
        connectionConfig = {
            host,
            port,
            database: 'test',
            user: 'postgres',
            password: 'postgres'
        };
        
        console.log(`PostgreSQL container started on ${host}:${port}`);
        
        // Setup test users with JWT tokens as passwords
        await setupEntraTestUsers(connectionConfig);
    });
    
    after(async function() {
        if (container) {
            console.log('Stopping PostgreSQL container...');
            await container.stop();
        }
    });
    
    /**
     * Setup test users with JWT tokens as passwords
     */
    async function setupEntraTestUsers(config) {
        const testUserToken = createValidJwtToken('test@example.com');
        const managedIdentityToken = createJwtTokenWithAppId('managed-identity-app-id');
        const fallbackUserToken = createValidJwtToken('fallback@example.com');
        
        const setupCommands = [
            `CREATE USER "test@example.com" WITH PASSWORD '${testUserToken}';`,
            `CREATE USER "managed-identity-app-id" WITH PASSWORD '${managedIdentityToken}';`,
            `CREATE USER "fallback@example.com" WITH PASSWORD '${fallbackUserToken}';`,
            'GRANT CONNECT ON DATABASE test TO "test@example.com";',
            'GRANT CONNECT ON DATABASE test TO "managed-identity-app-id";',
            'GRANT CONNECT ON DATABASE test TO "fallback@example.com";',
            'GRANT ALL PRIVILEGES ON DATABASE test TO "test@example.com";',
            'GRANT ALL PRIVILEGES ON DATABASE test TO "managed-identity-app-id";',
            'GRANT ALL PRIVILEGES ON DATABASE test TO "fallback@example.com";',
            'GRANT ALL ON SCHEMA public TO "test@example.com";',
            'GRANT ALL ON SCHEMA public TO "managed-identity-app-id";',
            'GRANT ALL ON SCHEMA public TO "fallback@example.com";',
            'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "test@example.com";',
            'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "managed-identity-app-id";',
            'GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO "fallback@example.com";',
            'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "test@example.com";',
            'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "managed-identity-app-id";',
            'GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO "fallback@example.com";'
        ];
        
        const client = new Client(config);
        await client.connect();
        
        for (const sql of setupCommands) {
            try {
                await client.query(sql);
            } catch (error) {
                // Ignore errors if user already exists
            }
        }
        
        await client.end();
    }
    
    /**
     * Helper to test Sequelize Entra connection works end-to-end
     */
    async function assertSequelizeEntraWorks(baseConfig, token, expectedUsername) {
        const credential = new TestTokenCredential(token);
        
        // Create Sequelize instance without username/password
        const sequelize = new Sequelize({
            ...baseConfig,
            logging: false
        });
        
        // Configure Entra authentication
        configureEntraIdAuth(sequelize, credential);
        
        // Test connection
        await sequelize.authenticate();
        
        // Verify we're connected as the expected user
        const [results] = await sequelize.query('SELECT current_user, current_database()');
        const { current_user, current_database } = results[0];
        
        expect(current_user).to.equal(expectedUsername);
        expect(current_database).to.equal('test');
        
        await sequelize.close();
    }
    
    /**
     * Helper to test node-postgres Entra connection works end-to-end
     */
    async function assertPgEntraWorks(baseConfig, token, expectedUsername) {
        const credential = new TestTokenCredential(token);
        
        // Get token
        const accessToken = await getEntraTokenPassword(credential);
        
        // Decode token to get username
        const claims = decodeJwtToken(accessToken);
        const username = claims.upn || claims.appid;
        
        // Create client with Entra credentials
        const client = new Client({
            ...baseConfig,
            user: username,
            password: accessToken
        });
        
        await client.connect();
        
        // Verify connection
        const result = await client.query('SELECT current_user, current_database()');
        const { current_user, current_database } = result.rows[0];
        
        expect(current_user).to.equal(expectedUsername);
        expect(current_database).to.equal('test');
        
        await client.end();
    }
    
    /**
     * Decode JWT token to extract user information
     */
    function decodeJwtToken(token) {
        try {
            const parts = token.split('.');
            if (parts.length !== 3) {
                throw new Error('Invalid JWT token format');
            }
            
            const payload = parts[1];
            const paddedPayload = payload + '='.repeat((4 - payload.length % 4) % 4);
            const decodedPayload = Buffer.from(paddedPayload, 'base64url').toString('utf8');
            
            return JSON.parse(decodedPayload);
        } catch (error) {
            console.error('Error decoding JWT token:', error);
            return null;
        }
    }
    
    describe('Sequelize Integration', function() {
        it('should connect with Entra user using configureEntraIdAuth', async function() {
            const baseConfig = {
                dialect: 'postgres',
                host: connectionConfig.host,
                port: connectionConfig.port,
                database: connectionConfig.database
            };
            
            const testToken = createValidJwtToken('test@example.com');
            await assertSequelizeEntraWorks(baseConfig, testToken, 'test@example.com');
        });
        
        it('should connect with managed identity using configureEntraIdAuth', async function() {
            const baseConfig = {
                dialect: 'postgres',
                host: connectionConfig.host,
                port: connectionConfig.port,
                database: connectionConfig.database
            };
            
            const miToken = createJwtTokenWithAppId('managed-identity-app-id');
            await assertSequelizeEntraWorks(baseConfig, miToken, 'managed-identity-app-id');
        });
    });
    
    describe('node-postgres Integration', function() {
        it('should connect with Entra user using getEntraTokenPassword', async function() {
            const baseConfig = {
                host: connectionConfig.host,
                port: connectionConfig.port,
                database: connectionConfig.database
            };
            
            const testToken = createValidJwtToken('test@example.com');
            await assertPgEntraWorks(baseConfig, testToken, 'test@example.com');
        });
        
        it('should connect with managed identity using getEntraTokenPassword', async function() {
            const baseConfig = {
                host: connectionConfig.host,
                port: connectionConfig.port,
                database: connectionConfig.database
            };
            
            const miToken = createJwtTokenWithAppId('managed-identity-app-id');
            await assertPgEntraWorks(baseConfig, miToken, 'managed-identity-app-id');
        });
    });
});
